lib/vector-core/src/event/vrl_target.rs

```diff
@@ -0,0 +1,750 @@
+use super::{Event, EventMetadata, LogEvent, Metric, MetricKind, Value};
+use crate::config::log_schema;
+use lookup::LookupBuf;
+use snafu::Snafu;
+use std::{collections::BTreeMap, convert::TryFrom, iter::FromIterator};
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// Metrics aren't interested in paths that have a length longer than 3
+/// The longest path is 2, and we need to check that a third segment doesn't exist as we don't want
+/// fields such as `.tags.host.thing`.
+const MAX_METRIC_PATH_DEPTH: usize = 3;
+
+/// An adapter to turn `Event`s into `vrl_core::Target`s.
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
+    // LogEvent is essentially just a destructured `event::LogEvent`, but without the semantics
+    // that `fields` must always be a Map variant
+    LogEvent(Value, EventMetadata),
+    Metric(Metric),
+}
+
+impl VrlTarget {
+    pub fn new(event: Event) -> Self {
+        match event {
+            Event::Log(event) => {
+                let (fields, metadata) = event.into_parts();
+                VrlTarget::LogEvent(Value::Map(fields), metadata)
+            }
+            Event::Metric(event) => VrlTarget::Metric(event),
+        }
+    }
+
+    /// Turn the target back into events.
+    ///
+    /// This returns an iterator of events as one event can be turned into multiple by assigning an
+    /// array to `.` in VRL.
+    pub fn into_events(self) -> impl Iterator<Item = Event> {
+        match self {
+            VrlTarget::LogEvent(value, metadata) => {
+                Box::new(value_into_log_events(value, metadata)) as Box<dyn Iterator<Item = Event>>
+            }
+            VrlTarget::Metric(metric) => {
+                Box::new(std::iter::once(Event::Metric(metric))) as Box<dyn Iterator<Item = Event>>
+            }
+        }
+    }
+}
+
+impl vrl_core::Target for VrlTarget {
+    fn insert(&mut self, path: &LookupBuf, value: vrl_core::Value) -> Result<(), String> {
```

Comment:
Perhaps it's out of scope for this, but I wonder if instead of cloning the event at the start of the remap transform we could move the clone to here.

`VrlTarget` could be something like:

```rust
pub struct VrlTarget {
    retain_original: boolean,
    originalEvent: Option<BTreeMap<String, Value>>,
    event: enum {
        LogEvent(Value, EventMetadata),
        Metric(Metric),
    } 
}
```

Then here we could:

```rust
  if self.retain_original && self.original.is_none() {
    self.original = self.value.as_map().clone();
  }
```

This could have a significant boost for scripts that abort the transaction prior to modifying any fields..

Reply:
🤔 that's true. I feel like I'd prefer to punt on that for now since this maintains the current status quo. I think, ideally, the whole event would be "copy-on-write" so we wouldn't need to worry about the clone cost here.